File: typescript/src/internal/strict-json.ts
import { ERROR_CODES, ErrorCode, protocolError } from "./errors.js";
import { MAX_JSON_BYTES, MAX_JSON_DEPTH, MAX_JSON_NODES } from "./limits.js";

/** 可进入 channels 协议层的 JSON 值。 */
export type JSONValue = null | boolean | number | string | JSONValue[] | { [key: string]: JSONValue };

/** 严格 JSON object 类型。 */
export type JSONObject = { [key: string]: JSONValue };

const textEncoder = new TextEncoder();

/** 将 UTF-8 字节或 JS 字符串转换为严格解析文本并执行字节上限检查。 */
export function jsonText(input: string | Uint8Array): string {
  if (typeof input === "string") {
    if (textEncoder.encode(input).byteLength > MAX_JSON_BYTES) {
      throw protocolError(ERROR_CODES.MESSAGE_TOO_LARGE, "JSON 输入超过 1,048,000 字节");
    }
    rejectUnpairedSurrogates(input);
    return input;
  }
  const copy = new Uint8Array(input);
  if (copy.byteLength > MAX_JSON_BYTES) {
    throw protocolError(ERROR_CODES.MESSAGE_TOO_LARGE, "JSON 输入超过 1,048,000 字节");
  }
  try {
    // 保留 BOM 让语法层拒绝它；Go 严格入口同样不把 BOM 当作 JSON whitespace。
    return new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(copy);
  } catch (cause) {
    throw protocolError(ERROR_CODES.INVALID_JSON, "输入不是合法 UTF-8", cause);
  }
}

/** 严格解析完整 JSON，拒绝重复字段、非法 surrogate 和资源超限。 */
export function parseStrictJSON(input: string | Uint8Array): JSONValue {
  return new Parser(jsonText(input)).parse();
}

/** 检查 object 只包含允许字段。 */
export function requireObjectKeys(object: JSONObject, allowed: readonly string[]): void {
  const set = new Set(allowed);
  for (const key of Object.keys(object)) {
    if (!set.has(key)) {
      throw protocolError(ERROR_CODES.UNKNOWN_FIELD, `未知字段 ${JSON.stringify(key)}`);
    }
  }
}

/** 校验运行时构造值是只含指定字段的 object；额外字段返回 UNKNOWN_FIELD。 */
export function requireExactObjectKeys(value: unknown, allowed: readonly string[], label: string, invalidCode: ErrorCode = ERROR_CODES.INVALID_BODY): asserts value is Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw protocolError(invalidCode, `${label} 必须是 object`);
  }
  const object = value as Record<string, unknown>;
  const allowedSet = new Set(allowed);
  const keys = Object.keys(object);
  for (const key of keys) {
    if (!allowedSet.has(key)) throw protocolError(ERROR_CODES.UNKNOWN_FIELD, `${label} 存在未知字段 ${JSON.stringify(key)}`);
  }
  if (keys.length !== allowed.length) throw protocolError(invalidCode, `${label} 缺少必需字段`);
}

/** 读取必需字段，严格区分缺字段和 JSON null。 */
export function requireField(object: JSONObject, key: string): JSONValue {
  if (!Object.prototype.hasOwnProperty.call(object, key)) {
    throw protocolError(ERROR_CODES.INVALID_JSON, `缺少必需字段 ${key}`);
  }
  return object[key];
}

/** 判断 JSON object（数组不属于 object 分支）。 */
export function isJSONObject(value: JSONValue): value is JSONObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function rejectUnpairedSurrogates(value: string): void {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      if (index + 1 >= value.length) {
        throw protocolError(ERROR_CODES.INVALID_JSON, "文本包含孤立 high surrogate");
      }
      const next = value.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) {
        throw protocolError(ERROR_CODES.INVALID_JSON, "文本 high surrogate 缺少配对");
      }
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      throw protocolError(ERROR_CODES.INVALID_JSON, "文本包含孤立 low surrogate");
    }
  }
}

class Parser {
  private index = 0;
  private nodes = 0;

  constructor(private readonly text: string) {}

  parse(): JSONValue {
    this.skipWhitespace();
    if (this.index >= this.text.length) {
      throw protocolError(ERROR_CODES.INVALID_JSON, "JSON 为空");
    }
    const value = this.parseValue(0);
    this.skipWhitespace();
    if (this.index !== this.text.length) {
      throw protocolError(ERROR_CODES.INVALID_JSON, "JSON 后存在未解析内容");
    }
    return value;
  }

  private parseValue(depth: number): JSONValue {
    this.nodes += 1;
    if (this.nodes > MAX_JSON_NODES) {
      throw protocolError(ERROR_CODES.MESSAGE_TOO_LARGE, "JSON 节点数超过 100,000");
    }
    const character = this.text[this.index];
    if (character === "{") return this.parseObject(depth);
    if (character === "[") return this.parseArray(depth);
    if (character === '"') return this.parseString();
    if (character === "t") return this.parseLiteral("true", true);
    if (character === "f") return this.parseLiteral("false", false);
    if (character === "n") return this.parseLiteral("null", null);
    if (character === "-" || isDigit(character)) return this.parseNumber();
    throw protocolError(ERROR_CODES.INVALID_JSON, "JSON 值类型不合法");
  }

  private parseObject(depth: number): JSONObject {
    if (depth >= MAX_JSON_DEPTH) {
      throw protocolError(ERROR_CODES.MESSAGE_TOO_LARGE, "JSON 嵌套深度超过 64");
    }
    this.index += 1;
    const object: JSONObject = Object.create(null) as JSONObject;
    this.skipWhitespace();
    if (this.consume("}")) return object;
    while (true) {
      this.skipWhitespace();
      if (this.text[this.index] !== '"') {
        throw protocolError(ERROR_CODES.INVALID_JSON, "object key 必须是字符串");
      }
      const key = this.parseString();
      if (typeof key !== "string") throw protocolError(ERROR_CODES.INVALID_JSON, "object key 类型不合法");
      if (Object.prototype.hasOwnProperty.call(object, key)) {
        throw protocolError(ERROR_CODES.INVALID_JSON, "JSON object 存在重复字段");
      }
      this.skipWhitespace();
      if (!this.consume(":")) throw protocolError(ERROR_CODES.INVALID_JSON, "object key 后缺少冒号");
      this.skipWhitespace();
      object[key] = this.parseValue(depth + 1);
      this.skipWhitespace();
      if (this.consume("}")) return object;
      if (!this.consume(",")) throw protocolError(ERROR_CODES.INVALID_JSON, "object 字段之间缺少逗号或右括号");
    }
  }

  private parseArray(depth: number): JSONValue[] {
    if (depth >= MAX_JSON_DEPTH) {
      throw protocolError(ERROR_CODES.MESSAGE_TOO_LARGE, "JSON 嵌套深度超过 64");
    }
    this.index += 1;
    const array: JSONValue[] = [];
    this.skipWhitespace();
    if (this.consume("]")) return array;
    while (true) {
      this.skipWhitespace();
      array.push(this.parseValue(depth + 1));
      this.skipWhitespace();
      if (this.consume("]")) return array;
      if (!this.consume(",")) throw protocolError(ERROR_CODES.INVALID_JSON, "array 元素之间缺少逗号或右括号");
    }
  }

  private parseString(): string {
    const start = this.index;
    this.index += 1;
    while (this.index < this.text.length) {
      const code = this.text.charCodeAt(this.index);
      if (code === 0x22) {
        this.index += 1;
        const raw = this.text.slice(start, this.index);
        let value: unknown;
        try {
          value = JSON.parse(raw);
        } catch (cause) {
          throw protocolError(ERROR_CODES.INVALID_JSON, "字符串转义不合法", cause);
        }
        if (typeof value !== "string") throw protocolError(ERROR_CODES.INVALID_JSON, "字符串类型不合法");
        rejectUnpairedSurrogates(value);
        return value;
      }
      if (code < 0x20) throw protocolError(ERROR_CODES.INVALID_JSON, "字符串包含未转义控制字符");
      if (code === 0x5c) {
        this.index += 1;
        const escape = this.text[this.index];
        if (escape === "u") {
          const unit = this.parseUnicodeEscape();
          if (unit >= 0xdc00 && unit <= 0xdfff) {
            throw protocolError(ERROR_CODES.INVALID_JSON, "字符串包含孤立 low surrogate");
          }
          if (unit >= 0xd800 && unit <= 0xdbff) {
            if (this.text[this.index] !== "\\" || this.text[this.index + 1] !== "u") {
              throw protocolError(ERROR_CODES.INVALID_JSON, "字符串 high surrogate 缺少配对");
            }
            this.index += 2;
            const low = this.parseUnicodeEscapeDigits();
            if (low < 0xdc00 || low > 0xdfff) {
              throw protocolError(ERROR_CODES.INVALID_JSON, "字符串 surrogate 配对不合法");
            }
          }
        } else if (!["\"", "\\", "/", "b", "f", "n", "r", "t"].includes(escape)) {
          throw protocolError(ERROR_CODES.INVALID_JSON, "字符串包含未知转义");
        } else {
          this.index += 1;
        }
        continue;
      }
      if (code >= 0xd800 && code <= 0xdbff) {
        if (this.index + 1 >= this.text.length || this.text.charCodeAt(this.index + 1) < 0xdc00 || this.text.charCodeAt(this.index + 1) > 0xdfff) {
          throw protocolError(ERROR_CODES.INVALID_JSON, "字符串包含孤立 high surrogate");
        }
        this.index += 2;
      } else if (code >= 0xdc00 && code <= 0xdfff) {
        throw protocolError(ERROR_CODES.INVALID_JSON, "字符串包含孤立 low surrogate");
      } else {
        this.index += 1;
      }
    }
    throw protocolError(ERROR_CODES.INVALID_JSON, "字符串缺少结束引号");
  }

  private parseUnicodeEscape(): number {
    this.index += 1;
    return this.parseUnicodeEscapeDigits();
  }

  private parseUnicodeEscapeDigits(): number {
    const raw = this.text.slice(this.index, this.index + 4);
    if (raw.length !== 4 || !/^[0-9a-fA-F]{4}$/.test(raw)) {
      throw protocolError(ERROR_CODES.INVALID_JSON, "Unicode 转义不是四位十六进制");
    }
    this.index += 4;
    return Number.parseInt(raw, 16);
  }

  private parseLiteral(literal: string, value: JSONValue): JSONValue {
    if (this.text.slice(this.index, this.index + literal.length) !== literal) {
      throw protocolError(ERROR_CODES.INVALID_JSON, "JSON literal 不完整");
    }
    this.index += literal.length;
    return value;
  }

  private parseNumber(): number {
    const start = this.index;
    if (this.text[this.index] === "-") {
      this.index += 1;
      if (this.index >= this.text.length) throw protocolError(ERROR_CODES.INVALID_JSON, "JSON number 不完整");
    }
    if (this.text[this.index] === "0") {
      this.index += 1;
      if (isDigit(this.text[this.index])) throw protocolError(ERROR_CODES.INVALID_JSON, "JSON number 不允许前导零");
    } else {
      if (!isNonZeroDigit(this.text[this.index])) throw protocolError(ERROR_CODES.INVALID_JSON, "JSON number 整数部分不合法");
      while (isDigit(this.text[this.index])) this.index += 1;
    }
    if (this.text[this.index] === ".") {
      this.index += 1;
      const fractionStart = this.index;
      while (isDigit(this.text[this.index])) this.index += 1;
      if (fractionStart === this.index) throw protocolError(ERROR_CODES.INVALID_JSON, "JSON number 小数部分不合法");
    }
    if (this.text[this.index] === "e" || this.text[this.index] === "E") {
      this.index += 1;
      if (this.text[this.index] === "+" || this.text[this.index] === "-") this.index += 1;
      const exponentStart = this.index;
      while (isDigit(this.text[this.index])) this.index += 1;
      if (exponentStart === this.index) throw protocolError(ERROR_CODES.INVALID_JSON, "JSON number 指数不合法");
    }
    const raw = this.text.slice(start, this.index);
    const value = Number(raw);
    if (!Number.isFinite(value)) throw protocolError(ERROR_CODES.INVALID_JSON, "JSON number 不能表示为有限 IEEE 754 double");
    return value;
  }

  private skipWhitespace(): void {
    while ([" ", "\t", "\r", "\n"].includes(this.text[this.index])) this.index += 1;
  }

  private consume(character: string): boolean {
    if (this.text[this.index] === character) {
      this.index += 1;
      return true;
    }
    return false;
  }
}

function isDigit(value: string | undefined): boolean {
  return value !== undefined && value >= "0" && value <= "9";
}

function isNonZeroDigit(value: string | undefined): boolean {
  return value !== undefined && value >= "1" && value <= "9";
}

File: internal/strictjson/strictjson.go
// Package strictjson 提供带重复字段和资源上限检查的 JSON 解析器。
package strictjson

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strconv"
	"unicode/utf8"

	"github.com/bsv8/ChannelProtocol/internal/protocolerror"
)

// MaxJSONBytes 是单份 JSON 输入和构造结果允许的最大 UTF-8 字节数。
const MaxJSONBytes = 1_048_000

// MaxJSONDepth 是 JSON 容器允许的最大嵌套深度。
const MaxJSONDepth = 64

// MaxJSONNodes 是一份 JSON 文档允许的最大值节点数。
const MaxJSONNodes = 100_000

// JSONValue 是严格解析后可进入协议层的 JSON 值。
type JSONValue any

type parser struct {
	data  []byte
	i     int
	nodes int
}

// Parse 严格解析完整 UTF-8 JSON，拒绝重复字段、非法 surrogate 和资源超限。
func Parse(data []byte) (JSONValue, error) {
	if len(data) > MaxJSONBytes {
		return nil, protocolerror.New(protocolerror.MessageTooLarge, "JSON 输入超过 1,048,000 字节")
	}
	if !utf8.Valid(data) {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "输入不是合法 UTF-8")
	}
	p := &parser{data: data}
	p.skipWhitespace()
	if p.i == len(p.data) {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON 为空")
	}
	value, err := p.parseValue(0)
	if err != nil {
		return nil, err
	}
	p.skipWhitespace()
	if p.i != len(p.data) {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON 后存在未解析内容")
	}
	return value, nil
}

// ParseObject 严格解析并要求根值为 object。
func ParseObject(data []byte) (map[string]JSONValue, error) {
	value, err := Parse(data)
	if err != nil {
		return nil, err
	}
	object, ok := value.(map[string]JSONValue)
	if !ok {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "协议根值必须是 object")
	}
	return object, nil
}

func (p *parser) parseValue(depth int) (JSONValue, error) {
	if p.nodes >= MaxJSONNodes {
		return nil, protocolerror.New(protocolerror.MessageTooLarge, "JSON 节点数超过 100,000")
	}
	p.nodes++
	if p.i >= len(p.data) {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON 意外结束")
	}
	switch p.data[p.i] {
	case '{':
		return p.parseObject(depth)
	case '[':
		return p.parseArray(depth)
	case '"':
		return p.parseString()
	case 't':
		return p.parseLiteral("true", true)
	case 'f':
		return p.parseLiteral("false", false)
	case 'n':
		return p.parseLiteral("null", nil)
	default:
		if p.data[p.i] == '-' || (p.data[p.i] >= '0' && p.data[p.i] <= '9') {
			return p.parseNumber()
		}
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON 值类型不合法")
	}
}

func (p *parser) parseObject(depth int) (JSONValue, error) {
	if depth >= MaxJSONDepth {
		return nil, protocolerror.New(protocolerror.MessageTooLarge, "JSON 嵌套深度超过 64")
	}
	p.i++ // {
	object := make(map[string]JSONValue)
	p.skipWhitespace()
	if p.consume('}') {
		return object, nil
	}
	for {
		p.skipWhitespace()
		if p.i >= len(p.data) || p.data[p.i] != '"' {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "object key 必须是字符串")
		}
		keyValue, err := p.parseString()
		if err != nil {
			return nil, err
		}
		key := keyValue.(string)
		if _, exists := object[key]; exists {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON object 存在重复字段")
		}
		p.skipWhitespace()
		if !p.consume(':') {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "object key 后缺少冒号")
		}
		p.skipWhitespace()
		value, err := p.parseValue(depth + 1)
		if err != nil {
			return nil, err
		}
		object[key] = value
		p.skipWhitespace()
		if p.consume('}') {
			return object, nil
		}
		if !p.consume(',') {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "object 字段之间缺少逗号或右括号")
		}
	}
}

func (p *parser) parseArray(depth int) (JSONValue, error) {
	if depth >= MaxJSONDepth {
		return nil, protocolerror.New(protocolerror.MessageTooLarge, "JSON 嵌套深度超过 64")
	}
	p.i++ // [
	array := make([]JSONValue, 0)
	p.skipWhitespace()
	if p.consume(']') {
		return array, nil
	}
	for {
		p.skipWhitespace()
		value, err := p.parseValue(depth + 1)
		if err != nil {
			return nil, err
		}
		array = append(array, value)
		p.skipWhitespace()
		if p.consume(']') {
			return array, nil
		}
		if !p.consume(',') {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "array 元素之间缺少逗号或右括号")
		}
	}
}

func (p *parser) parseString() (JSONValue, error) {
	start := p.i
	p.i++ // opening quote
	for p.i < len(p.data) {
		c := p.data[p.i]
		switch c {
		case '"':
			p.i++
			var result string
			if err := json.Unmarshal(p.data[start:p.i], &result); err != nil {
				return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串转义不合法")
			}
			return result, nil
		case '\\':
			p.i++
			if p.i >= len(p.data) {
				return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串转义意外结束")
			}
			switch p.data[p.i] {
			case '"', '\\', '/', 'b', 'f', 'n', 'r', 't':
				p.i++
			case 'u':
				unit, err := p.parseUnicodeEscape()
				if err != nil {
					return nil, err
				}
				if unit >= 0xDC00 && unit <= 0xDFFF {
					return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串包含孤立 low surrogate")
				}
				if unit >= 0xD800 && unit <= 0xDBFF {
					if p.i+1 >= len(p.data) || p.data[p.i] != '\\' || p.data[p.i+1] != 'u' {
						return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串 high surrogate 缺少配对")
					}
					p.i += 2
					low, err := p.parseUnicodeEscapeDigits()
					if err != nil {
						return nil, err
					}
					if low < 0xDC00 || low > 0xDFFF {
						return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串 surrogate 配对不合法")
					}
				}
			default:
				return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串包含未知转义")
			}
		default:
			if c < 0x20 {
				return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串包含未转义控制字符")
			}
			p.i++
		}
	}
	return nil, protocolerror.New(protocolerror.InvalidJSON, "字符串缺少结束引号")
}

func (p *parser) parseUnicodeEscape() (int, error) {
	p.i++ // u
	return p.parseUnicodeEscapeDigits()
}

func (p *parser) parseUnicodeEscapeDigits() (int, error) {
	if p.i+4 > len(p.data) {
		return 0, protocolerror.New(protocolerror.InvalidJSON, "Unicode 转义长度不足")
	}
	value, err := strconv.ParseUint(string(p.data[p.i:p.i+4]), 16, 16)
	if err != nil {
		return 0, protocolerror.New(protocolerror.InvalidJSON, "Unicode 转义不是十六进制")
	}
	p.i += 4
	return int(value), nil
}

func (p *parser) parseLiteral(literal string, value JSONValue) (JSONValue, error) {
	if p.i+len(literal) > len(p.data) || string(p.data[p.i:p.i+len(literal)]) != literal {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON literal 不完整")
	}
	p.i += len(literal)
	return value, nil
}

func (p *parser) parseNumber() (JSONValue, error) {
	start := p.i
	if p.consume('-') && p.i >= len(p.data) {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON number 不完整")
	}
	if p.i < len(p.data) && p.data[p.i] == '0' {
		p.i++
		if p.i < len(p.data) && p.data[p.i] >= '0' && p.data[p.i] <= '9' {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON number 不允许前导零")
		}
	} else {
		if p.i >= len(p.data) || p.data[p.i] < '1' || p.data[p.i] > '9' {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON number 整数部分不合法")
		}
		for p.i < len(p.data) && p.data[p.i] >= '0' && p.data[p.i] <= '9' {
			p.i++
		}
	}
	if p.i < len(p.data) && p.data[p.i] == '.' {
		p.i++
		fractionStart := p.i
		for p.i < len(p.data) && p.data[p.i] >= '0' && p.data[p.i] <= '9' {
			p.i++
		}
		if fractionStart == p.i {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON number 小数部分不合法")
		}
	}
	if p.i < len(p.data) && (p.data[p.i] == 'e' || p.data[p.i] == 'E') {
		p.i++
		if p.i < len(p.data) && (p.data[p.i] == '+' || p.data[p.i] == '-') {
			p.i++
		}
		exponentStart := p.i
		for p.i < len(p.data) && p.data[p.i] >= '0' && p.data[p.i] <= '9' {
			p.i++
		}
		if exponentStart == p.i {
			return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON number 指数不合法")
		}
	}
	raw := string(p.data[start:p.i])
	f, err := strconv.ParseFloat(raw, 64)
	// strconv 对下溢到有限 0 也返回 ErrRange；ECMAScript Number/JCS 接受该值。
	// 真正溢出到 Infinity 仍必须拒绝，避免 Go/TypeScript 数字语义分叉。
	if (err != nil && !errors.Is(err, strconv.ErrRange)) || math.IsInf(f, 0) || math.IsNaN(f) {
		return nil, protocolerror.New(protocolerror.InvalidJSON, "JSON number 不能表示为有限 IEEE 754 double")
	}
	return json.Number(raw), nil
}

func (p *parser) skipWhitespace() {
	for p.i < len(p.data) {
		switch p.data[p.i] {
		case ' ', '\t', '\r', '\n':
			p.i++
		default:
			return
		}
	}
}

func (p *parser) consume(expected byte) bool {
	if p.i < len(p.data) && p.data[p.i] == expected {
		p.i++
		return true
	}
	return false
}

// RequireObjectKeys 检查 object 只包含 expected 中定义的字段，并返回缺失字段。
func RequireObjectKeys(object map[string]JSONValue, expected ...string) error {
	allowed := make(map[string]struct{}, len(expected))
	for _, key := range expected {
		allowed[key] = struct{}{}
	}
	for key := range object {
		if _, ok := allowed[key]; !ok {
			return protocolerror.New(protocolerror.UnknownField, fmt.Sprintf("未知字段 %q", key))
		}
	}
	return nil
}

// RequireField 返回字段值，并严格区分缺字段与字段值为 null。
func RequireField(object map[string]JSONValue, key string) (JSONValue, error) {
	value, ok := object[key]
	if !ok {
		return nil, protocolerror.New(protocolerror.InvalidJSON, fmt.Sprintf("缺少必需字段 %q", key))
	}
	return value, nil
}

// IsNull 判断 JSON null。
func IsNull(value JSONValue) bool { return value == nil }

// CloneBytes 返回输入字节的防御性复制。
func CloneBytes(input []byte) []byte { return bytes.Clone(input) }
